File: 125_algorithms/_exercises/templates/_algorithms_challenges/leetcode/leetCode/Stack/232_ImplementQueueUsingStacks.cpp
/*
 * @Author: xuezaigds@gmail.com
 * @Last Modified time: 2016-04-24 12:16:45
 */

c.. Queue {
public:
    // Push element x to the back of queue.
    void push(int x) {
        in_stack.push(x);
    }

    // Removes the element from in front of queue.
    void pop(void) {
        move();
        out_stack.pop();
    }

    // Get the front element.
    int peek(void) {
        move();
        r_ out_stack.top();
    }

    // Return whether the queue is empty.
    bool empty(void) {
        r_ in_stack.empty() && out_stack.empty();
    }

private:
    void move(){
        __(out_stack.empty()){
            _____(!in_stack.empty()){
                out_stack.push(in_stack.top());
                in_stack.pop();
            }
        }
    }
    stack<int> in_stack;
    stack<int> out_stack;
};

File: 125_algorithms/_exercises/templates/_algorithms_challenges/leetcode/leetCode/Stack/232_ImplementQueueUsingStacks.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-
# @Author: xuezaigds@gmail.com


c.. Queue o..
    """
    Use python list as the underlying data structure for stack.
    Add a "move()" method to simplify code: it moves all elements
    of the "inStack" to the "outStack" when the "outStack" is empty.
    """
    ___ __init__(self
        self.in_stack, self.out_stack   # list, []

    ___ push  x
        self.in_stack.append(x)

    ___ pop(self
        self.move()
        self.out_stack.pop()

    ___ peek(self
        self.move()
        r_ self.out_stack[-1]

    ___ empty(self
        r_ (n.. self.in_stack) and (n.. self.out_stack)

    ___ move(self
        __ n.. self.out_stack:
            _____ self.in_stack:
                self.out_stack.append(self.in_stack.pop())

'''
if __name__ == '__main__':
    q = Queue()
    q.push(2)
    q.push(3)
    q.push(4)
    print q.peek()
    q.pop()
    print q.peek()
    q.pop()
    q.pop()
    print q.empty()
'''
